crates/qasm3/src/circuit.rs

```diff
@@ -0,0 +1,333 @@
+            /// efficiency; we can use direct list methods to retrieve the bit instances, rather
+            /// than needing to indirect through the general `__getitem__` of the register, or
+            /// looking up the qubit instances on the circuit.
+            items: Py<PyList>,
+        }
+
+        impl PyRegister for $name {
+            /// Get an individual bit from the register.
+            fn bit(&self, py: Python, index: usize) -> PyResult<Py<PyAny>> {
+                // Unfortunately, `PyList::get_item_unchecked` isn't usable with the stable ABI.
+                self.items
+                    .as_ref(py)
+                    .get_item(index)
+                    .map(|item| item.into_py(py))
+            }
+
+            fn bit_list<'a>(&'a self, py: Python<'a>) -> &'a PyList {
+                self.items.as_ref(py)
+            }
+        }
+
+        impl ::pyo3::IntoPy<Py<PyAny>> for $name {
+            fn into_py(self, _py: Python) -> Py<PyAny> {
+                self.object
+            }
+        }
+
+        impl ::pyo3::ToPyObject for $name {
+            fn to_object(&self, py: Python) -> Py<PyAny> {
+                // _Technically_, allowing access this internal object can let the Rust-space
+                // wrapper get out-of-sync since we keep a direct handle to the list, but in
+                // practice, the field it's viewing is private and "inaccessible" from Python.
+                self.object.clone_ref(py)
+            }
+        }
+    };
+}
+
+register_type!(PyQuantumRegister);
+register_type!(PyClassicalRegister);
+
+/// Information received from Python space about how to construct a Python-space object to
+/// represent a given gate that might be declared.
+#[pyclass(module = "qiskit._qasm3", frozen, name = "CustomGate")]
+#[derive(Clone, Debug)]
+pub struct PyGate {
+    constructor: Py<PyAny>,
+    name: String,
+    num_params: usize,
+    num_qubits: usize,
+}
+
+impl PyGate {
+    pub fn new<T: IntoPy<Py<PyAny>>, S: AsRef<str>>(
+        py: Python,
+        constructor: T,
+        name: S,
+        num_params: usize,
+        num_qubits: usize,
+    ) -> Self {
+        Self {
+            constructor: constructor.into_py(py),
+            name: name.as_ref().to_owned(),
+            num_params,
+            num_qubits,
+        }
+    }
+
+    /// Construct a Python-space instance of the custom gate.
+    pub fn construct<A>(&self, py: Python, args: A) -> PyResult<Py<PyAny>>
+    where
+        A: IntoPy<Py<PyTuple>>,
+    {
+        let args = args.into_py(py);
+        let received_num_params = args.as_ref(py).len();
+        if received_num_params == self.num_params {
+            self.constructor.call1(py, args.as_ref(py))
+        } else {
+            Err(QASM3ImporterError::new_err(format!(
+                "internal error: wrong number of params for {} (got {}, expected {})",
+                &self.name, received_num_params, self.num_params
+            )))
```

Comment:
These errors should be found by the parser (oq3_semantics). In fact for gate calls, this is already implemented. If we want to check for errors here anyway, it might be a good idea to note in the message that you have found a bug in openqasm3_parser.

Reply:
They should be, but because it's possible to represent in the typing I have to cover the case here and convert it to a Python exception so it happens loudly, not quietly returning the wrong thing.

Reply:
That's fine, but if it's triggered, it still has found a bug in the parser.

Reply:
Well, not necessarily - it could be that this code screwed something up while collecting the parameters or passed the wrong thing to `PyGate::construct`.  I think it's typically best to avoid naming projects in error messages - you never really know how they'll trigger, especially in a year's time.